File: Packages/WhoopProtocol/Sources/WhoopProtocol/PpgHr.swift
import Foundation

/// Per-second heart rate derived from the WHOOP 5.0 type-47 **v26** optical PPG buffer (issue #156).
///
/// On v26-heavy stretches of a night the strap records the 24 Hz optical PPG waveform instead of the
/// v18 per-second summary that carries HR. The v26 records are real cardiac signal, so this pure
/// estimator recovers a per-second HR from them via windowed autocorrelation, keeping the biometric
/// timeline continuous through the gaps. It is HR-only — PPG carries NO body motion, so this fills HR
/// continuity, NOT actigraphy and NOT HRV (the contributor confirmed v26 gives no RMSSD).
///
/// Mirrors `tools/linux-capture/ppg_hr.py` (PR #162, Python side): 8 s autocorrelation window, a
/// 30–220 bpm search band (lags 6…48 at 24 Hz), linear detrend, normalised peak as confidence, and a
/// fundamental-period preference so the harmonic peaks at 2×/3× the true period don't report half/third
/// the real rate. Pure + Foundation-only so it is unit-testable from synthetic and captured waveforms.

public struct PpgHrSample: Equatable, Codable, Sendable {
    public let ts: Int          // wall-clock unix seconds (one estimate per second)
    public let bpm: Int         // PPG-derived heart rate, whole bpm — the same Int domain as the
                                // measured HRSample.bpm and the Android PpgHr.Estimate.bpm, so the
                                // stored value and type match across platforms (#219).
    public let conf: Double     // normalised autocorrelation peak behind `bpm` (0…1)
    public init(ts: Int, bpm: Int, conf: Double) {
        self.ts = ts; self.bpm = bpm; self.conf = conf
    }
}

public enum PpgHr {
    public static let sampleRateHz = 24          // v26 carries 24 samples per 1-second record
    public static let windowSeconds = 8          // autocorrelation window (stable at low HR)
    public static let hrLoBpm = 30.0
    public static let hrHiBpm = 220.0
    public static let minConfidence = 0.3        // reject a window whose best peak is weaker than this

    /// Linear-detrend a waveform: subtract the least-squares best-fit line to remove DC + baseline
    /// wander (slow respiration/perfusion drift) before the autocorrelation, so the pulse dominates.
    static func detrend(_ x: [Double]) -> [Double] {
        let n = x.count
        guard n > 1 else { return x.map { _ in 0 } }
        let nD = Double(n)
        // x-axis is the sample index 0…n-1; closed-form slope/intercept of the LS line.
        let sumI = nD * (nD - 1) / 2
        let sumI2 = (nD - 1) * nD * (2 * nD - 1) / 6
        var sumY = 0.0, sumIY = 0.0
        for (i, y) in x.enumerated() { let id = Double(i); sumY += y; sumIY += id * y }
        let denom = nD * sumI2 - sumI * sumI
        guard denom != 0 else {
            let mean = sumY / nD
            return x.map { $0 - mean }
        }
        let slope = (nD * sumIY - sumI * sumY) / denom
        let intercept = (sumY - slope * sumI) / nD
        return x.enumerated().map { (i, y) in y - (slope * Double(i) + intercept) }
    }

    /// Normalised autocorrelation of `x` at `lag` (0 when the signal is flat).
    static func acf(_ x: [Double], _ lag: Int) -> Double {
        let n = x.count - lag
        guard n > 0 else { return 0 }
        let mean = x.reduce(0, +) / Double(x.count)
        var den = 0.0
        for v in x { let d = v - mean; den += d * d }
        guard den != 0 else { return 0 }
        var num = 0.0
        for i in 0..<n { num += (x[i] - mean) * (x[i + lag] - mean) }
        return num / den
    }

    /// Subtract the record-synchronous (period = `fs`) component of `x` — the artifact a per-record DC
    /// step / phase reset injects. It autocorrelates strongly at lag = fs (= 60 bpm at 24 Hz), so a
    /// sub-60-bpm sleeper would otherwise SNAP to 60 because fundamental-period preference picks the
    /// smaller lag (#194, ryanbr).
    ///
    /// CRUCIAL: a true ~60-bpm pulse is ALSO period-fs (one cycle per record), so a blind column-mean
    /// subtraction would erase a real 60 bpm. The discriminator is the record BOUNDARY: an artifact
    /// (phase reset / DC step) is DISCONTINUOUS there, while a real pulse flows smoothly across it. So
    /// only de-artifact when the boundary first-difference is much larger than the within-record one;
    /// otherwise leave the signal untouched. Gated on ≥ 4 records so the comb is well estimated.
    static func removeRecordRateComponent(_ x: [Double], fs: Int) -> [Double] {
        let n = x.count
        guard fs > 1, n >= fs * 4 else { return x }
        var withinSum = 0.0, withinCount = 0, boundarySum = 0.0, boundaryCount = 0
        for i in 1..<n {
            let d = abs(x[i] - x[i - 1])
            if i % fs == 0 { boundarySum += d; boundaryCount += 1 } else { withinSum += d; withinCount += 1 }
        }
        guard withinCount > 0, boundaryCount > 0 else { return x }
        let within = withinSum / Double(withinCount)
        let boundary = boundarySum / Double(boundaryCount)
        // Smooth boundaries → a real pulse, not an artifact → leave it alone (preserves a true 60 bpm).
        guard within > 0, boundary > within * 3 else { return x }
        var colSum = [Double](repeating: 0, count: fs)
        var colCount = [Int](repeating: 0, count: fs)
        for i in 0..<n { let p = i % fs; colSum[p] += x[i]; colCount[p] += 1 }
        var colMean = [Double](repeating: 0, count: fs)
        for p in 0..<fs where colCount[p] > 0 { colMean[p] = colSum[p] / Double(colCount[p]) }
        return x.enumerated().map { (i, v) in v - colMean[i % fs] }
    }

    /// Estimate (bpm, confidence) from one PPG window via autocorrelation, or nil when the window is
    /// too short or no pulsatile peak clears `minConfidence` (flat/garbage PPG → no fabricated HR).
    public static func estimate(_ samples: [Int],
                                fs: Int = sampleRateHz,
                                loBpm: Double = hrLoBpm,
                                hiBpm: Double = hrHiBpm,
                                minConf: Double = minConfidence,
                                subLagInterp: Bool = false) -> (bpm: Double, conf: Double)? {
        guard samples.count >= fs * 3 else { return nil }   // need >= 3 s to resolve a low HR
        // De-artifact (#194) THEN linear-detrend, so the autocorrelation sees the pulse, not the
        // record-rate comb that would peg a low HR at 60 bpm.
        let x = detrend(removeRecordRateComponent(samples.map(Double.init), fs: fs))
        let fsD = Double(fs)
        let loLag = max(2, Int((fsD * 60 / hiBpm).rounded()))
        let hiLag = min(x.count - 2, Int((fsD * 60 / loBpm).rounded()))
        guard hiLag > loLag else { return nil }
        var vals = [Int: Double]()
        var peak = -Double.infinity
        for lag in loLag...hiLag {
            let v = acf(x, lag)
            vals[lag] = v
            if v > peak { peak = v }
        }
        guard peak >= minConf else { return nil }
        // Prefer the FUNDAMENTAL period: the smallest-lag local maximum that is nearly as strong as the
        // global peak. Autocorrelation also peaks at 2×/3× the true period (half/third HR); taking the
        // global max there would report half the real rate, so prefer the shortest prominent period.
        var bestLag: Int? = nil
        if loLag + 1 <= hiLag - 1 {
            for lag in (loLag + 1)...(hiLag - 1) {
                let v = vals[lag]!
                if v >= 0.85 * peak && v >= vals[lag - 1]! && v >= vals[lag + 1]! {
                    bestLag = lag
                    break
                }
            }
        }
        let lag = bestLag ?? (vals.max { $0.value < $1.value }!.key)
        // Sub-lag parabolic interpolation (Variant A, opt-in): refine the integer `lag` by fitting a
        // parabola to the ACF peak and its two neighbours, so a true HR sitting BETWEEN two integer lags is
        // not quantized (integer lags step ~16 bpm near 150 bpm, so the estimate can be off up to ~±8 bpm).
        // Guarded to interior lags; a non-concave fit (denom >= 0) or a clamp-defended one falls back to the
        // integer lag. Default OFF is byte-identical to the integer-lag estimate. Mirror of the Kotlin branch.
        var refinedLag = Double(lag)
        if subLagInterp, lag - 1 >= loLag, lag + 1 <= hiLag {
            let y0 = vals[lag - 1]!, y1 = vals[lag]!, y2 = vals[lag + 1]!
            let denom = y0 - 2 * y1 + y2
            if denom < 0 {
                let delta = min(1.0, max(-1.0, 0.5 * (y0 - y2) / denom))
                refinedLag = min(Double(hiLag), max(Double(loLag), Double(lag) + delta))
            }
        }
        // Round to whole bpm (matching Android's Math.round) — whole bpm keeps the value + type in lockstep
        // with Android (#219). conf stays the integer `lag`'s peak (refinement moves frequency, not confidence).
        let bpm = (fsD * 60 / refinedLag).rounded()
        let conf = (vals[lag]! * 1000).rounded() / 1000
        return (bpm, conf)
    }

    /// Per-second PPG-HR over a list of v26 records `(ts, samples)`.
    ///
    /// Records are grouped into consecutive-second runs (PPG phase is only continuous within a run); a
    /// window centred on each second is autocorrelated. Returns one `PpgHrSample` per second that
    /// yielded a confident estimate, ascending by ts. Records may be unsorted / contain gaps.
    public static func derivePpgHr(records: [(ts: Int, samples: [Int])],
                                   fs: Int = sampleRateHz,
                                   windowSeconds: Int = windowSeconds,
                                   subLagInterp: Bool = false) -> [PpgHrSample] {
        guard !records.isEmpty else { return [] }
        // One waveform per second (last write wins on a duplicate ts).
        var secs = [Int: [Int]]()
        for r in records { secs[r.ts] = r.samples }
        let order = secs.keys.sorted()
        // Split into consecutive-second runs.
        var runs = [[Int]]()
        var cur = [order[0]]
        for u in order.dropFirst() {
            if u - cur.last! == 1 { cur.append(u) }
            else { runs.append(cur); cur = [u] }
        }
        runs.append(cur)

        let half = windowSeconds / 2
        var out = [PpgHrSample]()
        for run in runs where run.count >= 3 {
            let runSet = Set(run)
            for t in run {
                // Window of consecutive seconds present in this run, centred on t.
                var win = [Int]()
                for u in (t - half)...(t + half) where runSet.contains(u) { win.append(u) }
                guard win.count >= 3 else { continue }
                var sig = [Int]()
                for u in win { sig.append(contentsOf: secs[u]!) }
                if let est = estimate(sig, fs: fs, subLagInterp: subLagInterp) {
                    out.append(PpgHrSample(ts: t, bpm: Int(est.bpm), conf: est.conf))
                }
            }
        }
        out.sort { $0.ts < $1.ts }
        return out
    }
}

File: android/app/src/main/java/com/noop/protocol/PpgHr.kt
package com.noop.protocol

/**
 * Derive heart rate from the WHOOP 5.0/MG **v26** optical PPG waveform (#156).
 *
 * The strap's type-47 **layout v26** record is a 24 Hz PPG buffer — NOT a per-second biometric
 * summary like v18: **24 little-endian i16 samples at frame bytes [27:75]**, one record per second,
 * with the record's own unix u32 LE @15 (the same slot v18 uses). WHOOP does NOT store a per-second
 * HR in v26 — HR is PPG-derived on-device — so to recover HR we re-derive it here from the waveform,
 * **byte-for-byte mirroring the Swift estimator** (WhoopProtocol/PpgHr.swift) so macOS, iOS and
 * Android produce the SAME per-second HR from the same offload. (Provenance: the concatenated
 * waveform's autocorrelation peaks at the heart rate; verified against measured HR as internal
 * ground truth — lag 14 ≈ 102.9 bpm vs a measured 101.7 bpm.)
 *
 * Algorithm (kept in lockstep with the Swift lane — #219 parity audit):
 *   • Records are grouped into **consecutive-second runs** (PPG phase is only continuous within a
 *     run); a window of the seconds present in `t-half … t+half` (half = WINDOW_SECONDS/2) is
 *     autocorrelated, one estimate per second whose centred window holds ≥ 3 seconds.
 *   • Per window: **linear least-squares detrend** (removes DC *and* baseline wander), then
 *     normalised autocorrelation over the lag band that maps to **30…220 bpm**.
 *   • **Fundamental-period preference**: pick the smallest-lag local maximum that is ≥ 0.85× the
 *     global peak, so the harmonic peaks at 2×/3× the true period don't report half/third the real
 *     rate. Falls back to the global argmax when no clean local max is found.
 *   • Emit only when the global peak clears **0.3** — a clean pulse autocorrelates strongly, noise
 *     does not, so a noisy/limp window yields nothing rather than a fabricated bpm.
 *   • bpm = 60·fs/lag, rounded to whole; the estimate's timestamp is the window's CENTRE second.
 *
 * Pure + side-effect-free so it is unit-testable on synthetic signals (see PpgHrTest).
 */
object PpgHr {
    /** PPG sample rate of the v26 waveform (Hz). */
    const val SAMPLE_RATE_HZ = 24

    /** HR estimation window length in seconds (centred half-window = WINDOW_SECONDS/2 each side). */
    const val WINDOW_SECONDS = 8

    /** Physiological HR search bounds (bpm). */
    const val MIN_BPM = 30.0
    const val MAX_BPM = 220.0

    /** Minimum normalised-autocorrelation peak to emit an estimate. */
    const val MIN_CONFIDENCE = 0.3

    /**
     * One concatenated, time-ordered PPG sample: its wall-clock second [ts] and raw ADC [value].
     * Built from contiguous v26 records (each record contributes 24 samples spanning one second).
     */
    data class Sample(val ts: Long, val value: Int)

    /** A derived HR estimate: [ts] = window-centre second, [bpm], [conf] in 0…1. */
    data class Estimate(val ts: Long, val bpm: Int, val conf: Double)

    /**
     * Per-second PPG-HR over the concatenated [samples] (mirror of Swift `derivePpgHr`).
     *
     * [samples] carry one [ts] per strap-second (all 24 samples of a record share it). They may be
     * unsorted or contain gaps: records are grouped by second (last write wins on a duplicate ts),
     * split into consecutive-second runs, and a centred window is autocorrelated for each second.
     * Returns one [Estimate] per second that yielded a confident estimate, ascending by ts.
     */
    fun estimate(samples: List<Sample>, subLagInterp: Boolean = false): List<Estimate> {
        if (samples.isEmpty()) return emptyList()
        // One waveform per second, in first-seen sample order (last record wins on a duplicate ts).
        val secs = LinkedHashMap<Long, ArrayList<Int>>()
        for (s in samples) {
            val list = secs.getOrPut(s.ts) { ArrayList() }
            list.add(s.value)
        }
        val order = secs.keys.sorted()

        // Split into consecutive-second runs.
        val runs = ArrayList<ArrayList<Long>>()
        var cur = arrayListOf(order[0])
        for (i in 1 until order.size) {
            val u = order[i]
            if (u - cur.last() == 1L) {
                cur.add(u)
            } else {
                runs.add(cur)
                cur = arrayListOf(u)
            }
        }
        runs.add(cur)

        val half = WINDOW_SECONDS / 2
        val out = ArrayList<Estimate>()
        for (run in runs) {
            if (run.size < 3) continue
            val runSet = run.toHashSet()
            for (t in run) {
                // Window of consecutive seconds present in this run, centred on t.
                val win = ArrayList<Long>()
                var u = t - half
                while (u <= t + half) {
                    if (u in runSet) win.add(u)
                    u++
                }
                if (win.size < 3) continue
                var total = 0
                for (w in win) total += secs[w]!!.size
                val sig = DoubleArray(total)
                var idx = 0
                for (w in win) for (v in secs[w]!!) sig[idx++] = v.toDouble()
                estimateWindow(sig, t, subLagInterp)?.let { out.add(it) }
            }
        }
        out.sortBy { it.ts }
        return out
    }

    /**
     * Linear-detrend a waveform: subtract the least-squares best-fit line to remove DC + baseline
     * wander (slow respiration/perfusion drift) before the autocorrelation, so the pulse dominates.
     * Mirror of Swift `PpgHr.detrend`.
     */
    private fun detrend(x: DoubleArray): DoubleArray {
        val n = x.size
        if (n <= 1) return DoubleArray(n) // 0 for all (matches Swift's x.map { 0 })
        val nD = n.toDouble()
        val sumI = nD * (nD - 1) / 2
        val sumI2 = (nD - 1) * nD * (2 * nD - 1) / 6
        var sumY = 0.0
        var sumIY = 0.0
        for (i in 0 until n) { sumY += x[i]; sumIY += i.toDouble() * x[i] }
        val denom = nD * sumI2 - sumI * sumI
        if (denom == 0.0) {
            val mean = sumY / nD
            return DoubleArray(n) { x[it] - mean }
        }
        val slope = (nD * sumIY - sumI * sumY) / denom
        val intercept = (sumY - slope * sumI) / nD
        return DoubleArray(n) { x[it] - (slope * it.toDouble() + intercept) }
    }

    /** Normalised autocorrelation of [x] at [lag] (0 when the signal is flat). Mirror of Swift `acf`. */
    private fun acf(x: DoubleArray, lag: Int): Double {
        val n = x.size - lag
        if (n <= 0) return 0.0
        var mean = 0.0
        for (v in x) mean += v
        mean /= x.size
        var den = 0.0
        for (v in x) { val d = v - mean; den += d * d }
        if (den == 0.0) return 0.0
        var num = 0.0
        for (i in 0 until n) num += (x[i] - mean) * (x[i + lag] - mean)
        return num / den
    }

    /**
     * Estimate HR for one window via linear detrend + normalised autocorrelation with
     * fundamental-period preference. Returns null when the window is too short (< 3 s), flat, or no
     * lag clears [MIN_CONFIDENCE]. Mirror of Swift `PpgHr.estimate` wrapped with the centre [ts].
     *
     * Lag band: a faster HR is a SHORTER lag, so [MAX_BPM] → loLag and [MIN_BPM] → hiLag. Bounds use
     * round-to-nearest and clamp to [2, n-2] exactly as Swift does.
     */
    /**
     * Subtract the record-synchronous (period = fs) component — the artifact a per-record DC step /
     * phase reset injects, which autocorrelates at lag = fs (60 bpm) and would SNAP a sub-60-bpm
     * sleeper to 60 via the fundamental-period preference (#194, ryanbr). A true ~60-bpm pulse is also
     * period-fs, so we only de-artifact when the record BOUNDARY is discontinuous (the artifact's
     * signature) — a real pulse flows smoothly across it and is left untouched, preserving a true
     * 60 bpm. Mirror of the Swift PpgHr.removeRecordRateComponent.
     */
    private fun removeRecordRateComponent(x: DoubleArray, fs: Int): DoubleArray {
        val n = x.size
        if (fs <= 1 || n < fs * 4) return x
        var withinSum = 0.0; var withinCount = 0; var boundarySum = 0.0; var boundaryCount = 0
        for (i in 1 until n) {
            val d = Math.abs(x[i] - x[i - 1])
            if (i % fs == 0) { boundarySum += d; boundaryCount++ } else { withinSum += d; withinCount++ }
        }
        if (withinCount == 0 || boundaryCount == 0) return x
        val within = withinSum / withinCount
        val boundary = boundarySum / boundaryCount
        if (within <= 0.0 || boundary <= within * 3) return x // smooth → real pulse → leave it
        val colSum = DoubleArray(fs); val colCount = IntArray(fs)
        for (i in 0 until n) { val p = i % fs; colSum[p] += x[i]; colCount[p]++ }
        val colMean = DoubleArray(fs) { p -> if (colCount[p] > 0) colSum[p] / colCount[p] else 0.0 }
        return DoubleArray(n) { i -> x[i] - colMean[i % fs] }
    }

    private fun estimateWindow(values: DoubleArray, ts: Long, subLagInterp: Boolean = false): Estimate? {
        if (values.size < SAMPLE_RATE_HZ * 3) return null // need >= 3 s to resolve a low HR
        // De-artifact (#194) THEN detrend, so the autocorrelation sees the pulse, not a record-rate comb.
        val x = detrend(removeRecordRateComponent(values, SAMPLE_RATE_HZ))
        val fsD = SAMPLE_RATE_HZ.toDouble()
        val loLag = maxOf(2, Math.round(fsD * 60 / MAX_BPM).toInt())
        val hiLag = minOf(x.size - 2, Math.round(fsD * 60 / MIN_BPM).toInt())
        if (hiLag <= loLag) return null

        val vals = HashMap<Int, Double>(hiLag - loLag + 1)
        var peak = Double.NEGATIVE_INFINITY
        for (lag in loLag..hiLag) {
            val v = acf(x, lag)
            vals[lag] = v
            if (v > peak) peak = v
        }
        if (peak < MIN_CONFIDENCE) return null

        // Prefer the FUNDAMENTAL period: the smallest-lag local maximum that is nearly as strong as
        // the global peak. Autocorrelation also peaks at 2×/3× the true period (half/third HR); the
        // global max there would report half the real rate, so prefer the shortest prominent period.
        var bestLag = -1
        if (loLag + 1 <= hiLag - 1) {
            for (lag in (loLag + 1)..(hiLag - 1)) {
                val v = vals[lag]!!
                if (v >= 0.85 * peak && v >= vals[lag - 1]!! && v >= vals[lag + 1]!!) {
                    bestLag = lag
                    break
                }
            }
        }
        if (bestLag < 0) {
            // Fallback: global argmax, smallest lag wins a tie (deterministic).
            var argmax = loLag
            var best = vals[loLag]!!
            for (lag in (loLag + 1)..hiLag) {
                val v = vals[lag]!!
                if (v > best) { best = v; argmax = lag }
            }
            bestLag = argmax
        }

        // Sub-lag parabolic interpolation (Variant A, opt-in): refine the integer bestLag by fitting a
        // parabola to the ACF peak and its two neighbours, so a true HR sitting BETWEEN two integer lags is
        // not quantized (integer lags step ~16 bpm near 150 bpm, so the estimate can be off up to ~±8 bpm).
        // Guarded to interior lags; a non-concave fit (denom >= 0) or a clamp-defended one falls back to the
        // integer lag. Default OFF is byte-identical to the integer-lag estimate. Mirror of the Swift branch.
        var refinedLag = bestLag.toDouble()
        if (subLagInterp && bestLag - 1 >= loLag && bestLag + 1 <= hiLag) {
            val y0 = vals[bestLag - 1]!!
            val y1 = vals[bestLag]!!
            val y2 = vals[bestLag + 1]!!
            val denom = y0 - 2.0 * y1 + y2
            if (denom < 0.0) {
                val delta = (0.5 * (y0 - y2) / denom).coerceIn(-1.0, 1.0)
                refinedLag = (bestLag.toDouble() + delta).coerceIn(loLag.toDouble(), hiLag.toDouble())
            }
        }
        // Round to whole bpm (the ppgHrSample column is Int; #219) — conf stays the integer bestLag's peak
        // (the refinement moves only the frequency, not the confidence). Matches the Swift estimator.
        val bpm = Math.round(fsD * 60 / refinedLag).toInt()
        val conf = Math.round(vals[bestLag]!! * 1000) / 1000.0
        return Estimate(ts = ts, bpm = bpm, conf = conf)
    }
}
